shared/src/question-generators/propositional-logic/minimizeDNF.ts

```diff
@@ -0,0 +1,292 @@
+import { Literal, Operator, SyntaxTreeNodeType, TruthTable } from "@shared/utils/propositionalLogic"
+
+/**
+ * During the Quine–McCluskey algorithm, the datatype for a truthTable can be a BinaryMix
+ * 0 --> false; 1 --> true; x --> true/false
+ */
+type BinaryMix = 0 | 1 | "x"
+
+/**
+ * Based on a given expression (truthTable), calculates a minimal DNF
+ * Using **Quine–McCluskey** algorithm
+ *
+ * @param expr
+ */
+export function minimizeExpressionDNF(expr: SyntaxTreeNodeType): SyntaxTreeNodeType {
+  const { truthTable: tTable, variableNames: varNames } = expr.getTruthTable()
+
+  const trueRows = getIndexOfTrueRows(tTable)
+  const { finalImplicants } = getPrimeImplicants(trueRows, varNames.length)
+  const { implicantRows } = getImplicantRows(finalImplicants)
+  const minRows = minimumCover(implicantRows, trueRows)
+
+  return buildMinDNFExpression(minRows, finalImplicants, varNames)
+}
+
+/**
+ * Creates the SyntaxTreeNode for the minimized dnf
+ * - each row in minRows is one clause
+ * - finalImplicants[row] -> [x0, x1, ..., xn] xi in {0, 1, x} says how to create each literal
+ *  - if 0: negated variable
+ *  - if 1: normal variable
+ *  - if x: don't add (this clause does not rely on the variable)
+ * @param minRows
+ * @param finalImplicants
+ * @param varNames
+ */
+function buildMinDNFExpression(minRows: any[], finalImplicants: BinaryMix[][], varNames: string[]) {
```

Comment:
ok, I looked into that in more detail. Using the Quine–McCluskey algorithm seems a good choice. On a first glance at the code it looked like the algorithm optimizes a given normal form. But this is not the case. However, this algorithm should be part of the propositional logic classes. Also have a look at the normal form functions. they are written in a generic way, that they produce DNF and CNF by only changing a lambda or a flag or so, to remove code duplication.

Reply:
As discussed I wrote a class for this algorithm and made some little other changes. Also minimally prepared the structure for CNF option. 

Since I don't know the algorithm for CNF yet many functions are not in the abstract class, instead in the DNF. But could be refactored into the abstract if the CNF algorithm uses those functions too